File: app/src/main/java/com/juliuscanute/multiconfig/ui/config/ConfigurationViewModel.kt
package com.juliuscanute.multiconfig.ui.config

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.juliuscanute.multiconfig.builder.ApplicationConfiguration
import com.juliuscanute.multiconfig.base.SingleObserverEvent
import com.juliuscanute.multiconfig.ui.adapter.ConfigurationViewDataModel
import com.juliuscanute.multiconfig.utils.ConfigManagerInitializer
import com.juliuscanute.multiconfig.model.ConfigurationManager

class ConfigurationViewModel : ViewModel() {
    private val configManager: ConfigurationManager by ConfigManagerInitializer()
    private val privateState = MutableLiveData<SingleObserverEvent<ConfigurationState>>()

    val privateActions: LiveData<SingleObserverEvent<ConfigurationState>> = privateState

    fun moveToConfigurationDetail(environment: String) {
        privateState.postValue(
            SingleObserverEvent(
                ConfigurationState.SelectedConfigurationState(
                    environment = environment
                )
            )
        )
    }

    fun selectNewConfiguration(selected: Int): Boolean {
        configManager.saveConfig(selected)
        loadApplicationConfiguration()
        return true
    }

    fun loadApplicationConfiguration() {
        val applicationConfiguration = configManager.getApplicationConfiguration()
        val selectedConfig = configManager.getConfig()
        val selectedIndex = if (selectedConfig < 0) 0 else selectedConfig
        privateState.postValue(
            SingleObserverEvent(
                ConfigurationState.LoadApplicationConfigurationState(
                    applicationConfiguration.mapState(selectedIndex),
                    applicationConfiguration[selectedConfig].environment
                )
            )
        )
    }
}

fun ApplicationConfiguration.mapState(selectedIndex: Int = 0): List<ConfigurationViewDataModel> {
    return this.mapIndexed { index, configuration ->
        if (selectedIndex == index)
            ConfigurationViewDataModel(
                index = index,
                environment = configuration.environment,
                selected = true
            )
        else
            ConfigurationViewDataModel(
                index = index,
                environment = configuration.environment,
                selected = false
            )
    }
}

File: iosApp/iosApp/ConfigurationViewModel.swift
//
//  ConfigurationViewModel.swift
//  iosApp
//
//  Created by Julius Canute on 29/3/20.
//

import Foundation
import app

public class ConfigurationViewModel {
    let configManager: ConfigurationManager
    let state = Box<ConfigurationState?>(nil)
    
    init(manager: () -> ConfigurationManager?) {
        self.configManager = manager()!
    }
    
    func moveToConfigurationDetail(environment: String) {
        
    }
    
    func selectNewConfiguration(selected: Int) {
        configManager.saveConfig(value: Int32(selected))
        loadApplicationConfiguration()
    }
    
    func loadApplicationConfiguration() {
        let applicationConfiguration: ApplicationConfiguration = configManager.getApplicationConfiguration() as! ApplicationConfiguration
        let selectedConfig = configManager.getConfig()
        let selectedIndex = selectedConfig < 0 ? 0 : selectedConfig
        let appState = LoadApplicationConfigurationState(items: applicationConfiguration.mapState(selectedIndex: Int(selectedIndex)), environment: applicationConfiguration[Int(selectedIndex)].environment, selectedIndex: Int(selectedIndex))
        state.value = ConfigurationState.appConfig(appState)
    }
}

extension ApplicationConfiguration {
    func mapState(selectedIndex: Int = 0) -> [ConfigurationViewDataModel] {
        return self.enumerated().map{ (index: Int, configuration: Configuration) in
            if(selectedIndex == index) {
                return ConfigurationViewDataModel(index: index, environment: configuration.environment, selected: true)
            } else {
                return ConfigurationViewDataModel(index: index, environment: configuration.environment, selected: false)
            }
        }
    }
}
